src/lib.rs

```diff
@@ -277,3 +285,139 @@ where
     /// The pairs `(A, Message)` indicate from which address each packet was received.
     fn receive_all_messages(&mut self) -> Vec<(A, Message)>;
 }
+
+/// An [InputPredictor] allows GGRS to predict the next input for a player based on previous input
+/// received.
+///
+/// A correct prediction means a rollback will not happen when input is received late from a remote
+/// player. An incorrect prediction will cause GGRS to request your game to rollback. It is normal
+/// and expected that some predictions will be incorrect, but the more incorrect predictions are
+/// given to GGRS, the more work your game will have to do to resimulate the game state (and the
+/// more that rollbacks may be noticeable to your human players).
+///
+/// For example, if your chosen input predictor says a player's input always makes them crouch, but
+/// in your game players only crouch in 1% of frames, then:
+///
+/// * GGRS will make it seem to your game as if all remote players crouch on every frame.
+/// * When GGRS receives input from a remote player and finds out they are not crouching, it will
+///   ask your game to rollback to the frame that input was from and resimulate it plus all
+///   subsequent frames up to and including the present frame.
+/// * Therefore 99% of frames will be resimulated.
+///
+/// [RepeatLastInputPredictor] is a good default choice for most games; other bundled predictors
+/// include [ZeroedInputPredictor] and [DefaultInputPredictor].
+///
+/// You are welcome to implement your own predictor, which is useful in cases where you are quite
+pub trait InputPredictor<I> {
+    /// Predict the next input for a player based on a possibly-available previous input.
+    ///
+    /// If this returns `None`, GGRS will attempt to manufacture a suitable fallback input.
+    /// Currently GGRS will use the zeroed out form of the input (see Bytemuck's
+    /// [zeroed()](Zeroable::zeroed)) as the fallback.
+    ///
+    /// The previous input may not be available, for example in the case where no input from a
+    /// remote player has been received in this session yet (notably, the very first simulation of
+    /// the first frame of a session will never have any inputs from remote players). In such a case
+    /// it is typically good to return `None`, `Some(I::default())`, or a Some with a value of `I`
+    /// that means "this player sent no input" for your game.
+    ///
+    fn predict(previous: Option<I>) -> Option<I>;
```

Comment:
What's the benefit of having this return an `Option`? Couldn't you just choose a `ZeroedInputPredictor` instead? or zero the input yourself if a previous input is not available?

Also is only the previous input enough? Maybe you'd want to estimate the velocity of change for an input and extrapolate?

Reply:
> What's the benefit of having this return an Option?

Yep, see point `C` in PR body (as per your edit to your parent comment); in hindsight I agree it's silly and would be better to not allow returning `Option`.

> Also is only the previous input enough?

Partially discussed this too under point `D` in PR body; exposing more information is possible but I don't personally have a use case for it (and seems like nobody else has requested it?), so I'm wary of making the API unnecessarily complicated (or not useful) to support that - or worse, introducing a subtle bug, as the code in `InputQueue::input()` is already a bit fiddly.

> take a slice of `I` if possible

Conceptually that'd be reasonable, but the inputs are stored in a Vec-based ring buffer so we'd have to pass 2 slices or do a bit of copying to achieve that. A good example of how the implementation/API would get complicated by supporting such more advanced input predictors!

A less invasive (for ggrs) way to do it might be to pass the frame number for both the current prediction request and the last confirmed prediction over to the input predictor, plus the player handle; then the input predictor can store as much of that as it wants in some separate data store. I had partially implemented that at one point before I backed it out - totally doable but some more data needs to be passed to InputQueue::input() to make it happen - if anyone ever actually requests it, of course!